woocommerce/woocommerce-ios
# Track store snapshot when the merchant uses a store for the first time

Closes: #10481 


## Why


As per the data request pdND4G-gg-p2#add-tracks-event-to-the-apps-to-gather-a-snapshot-at-the-time-of-install, we want to track the snapshot of each site once after app installation. This helps us gather data before the WCTracker changes take effect since it requires updates on the WC plugin and mobile apps.

The snapshot includes (ref pdND4G-gg-p2#comment-244):

- The number of products the store has - `orders_count: number`: reuse the result of `wc/v3/reports/orders/totals`
- The number of orders the store has - `products_count: number`: this requires a new API request `wc/v3/reports/products/totals`
- The payment gateway that is currently active, focusing on 4 plugins:
  - `woocommerce-payments: string` with one of the 3 values *
  - `woocommerce-gateway-stripe: string` with one of the 3 values *
  - `woocommerce-square: string` with one of the 3 values *
  - `woocommerce-gateway-paypal-express-checkout: string` with one of the 3 values *

*: `not_installed/installed_and_not_activated/installed_and_activated` to reuse the result of `wc/v3/system_status`

## How

Apology for the 500+ changes since there is a new API request to load the number of products.

🗒️ Some changes on `Customer` were unrelated but required due to the Copiable generation script. The order of the parameters in the initializer needs to follow the property order.

### Networking layer

Added `ProductsRemote.loadNumberOfProducts` to get the number of products from the `wc/v3/reports/products/totals` endpoint.

### Yosemite layer

Added `ProductAction.fetchNumberOfProducts` to load the number of products remotely.

### App layer

Added `SiteSnapshotTracker` to track the snapshot event `application_store_snapshot` with all the required event properties from the data fetched in `DefaultStoresManager`. It checks if the site has been tracked in `UserDefaults`'s new key `siteIDsWithSnapshotTracked` that returns a list of tracked site IDs. Since the products total API request isn't needed when snapshot tracking isn't needed, the order statuses and system plugins API requests were made async beforehand.

## Testing instructions


- Delete the app if it's already installed
- Install the app from the PR branch
- Log in to an account with multiple stores
- Select a site --> shortly after entering the site, an event should be logged `🔵 Tracked application_store_snapshot, properties: [AnyHashable("woocommerce-gateway-stripe"): "installed_and_activated", AnyHashable("products_count"): 82, AnyHashable("orders_count"): 583, AnyHashable("woocommerce-square"): "installed_and_not_activated", AnyHashable("woocommerce-payments"): "not_installed", AnyHashable("woocommerce-gateway-paypal-express-checkout"): "installed_and_not_activated"]` where the values match the store (🗒️ the orders/products count can be a bit off from the numbers in core from my testing, discussion p1692174309965889/1692000036.800969-slack-C6H8C3G23)
- Go to the Menu tab and select a different store --> shortly after entering the site, an event should be logged `application_store_snapshot` with the 6 parameters where the values match the store
- Go to the Menu tab and select the first store again --> `application_store_snapshot` shouldn't be tracked
- Relaunch the app --> shortly after the launch, `application_store_snapshot` shouldn't be tracked

---
- [x] @jaclync tests logging in with application password

---
